src/blocks.js

```diff
@@ -479,95 +479,77 @@ export default class CodeMirrorBlocks {
     event.preventDefault();
     event.stopPropagation();
     event.target.classList.remove('blocks-over-target');
-    let nodeId = event.dataTransfer.getData('text/id');
+    // look up the source information: ID, text, JSON, and the node itself
+    let sourceId       = event.dataTransfer.getData('text/id');
     let sourceNodeText = event.dataTransfer.getData('text/plain');
     let sourceNodeJSON = event.dataTransfer.getData('text/json');
-    let sourceNode = null;
-
-    if (nodeId) {
-      sourceNode = this.ast.nodeMap.get(nodeId);
-      if (sourceNode) {
-        sourceNodeText = this.cm.getRange(sourceNode.from, sourceNode.to);
-      } else {
-        console.error("node", nodeId, "not found in AST");
-      }
+    let sourceNode     = this.ast.nodeMap.get(sourceId);
+    if (sourceNode) {
+      sourceNodeText = this.cm.getRange(sourceNode.from, sourceNode.to);
     } else if (sourceNodeJSON) {
       sourceNode = JSON.parse(sourceNodeJSON);
     } else if (!sourceNodeText) {
       console.error("data transfer contains no node id/json/text. Not sure how to proceed.");
     }
 
-    let destination = getLocationFromEl(event.target);
-
-    if (!destination) {
-      // event.target probably isn't a drop target, so just get the location from the event
-      destination = this.cm.coordsChar({left:event.pageX, top:event.pageY});
-      if (destination.outside) {
-        sourceNodeText = '\n' + sourceNodeText;
-      }
+    // look up the destination information: ID, Node, destFrom and destTo
+    let destinationId   = event.target.id.substring(11);            // remove "blocks-node-" from the front
```

Comment:
Ah I see. I think the semantics of `findNearestNodeEl` are correct. But you should at least extract out the common code for going from element &rarr; AST node based on the element id.